[PROD4POD-1523] multiple archives

timoteipalade opened the pull request:
https://jira.polypoly.eu/browse/PROD4POD-1523

To support multiple archives I want to modify polyOut.importArchive to support a path argument, the idea being that you can import multiple archives in the same place. 

So far I have done the implementation only for iOS and I would like some feedback. Thanks! 

The idea is that one will be able to import multiple archives in the same place like this: 
```
const destUrl = polyOut.importArchive(archiveUrl_1)
polyOut.importArchive(archiveUrl_2, destUrl)
```

JJ commented on platform/ios/PolyPodApp/PodApi/PolyOut/FS/PolyOut+FS.swift:
```diff
@@ -93,7 +93,7 @@ extension PolyOut {
                     time: "\(Int(floor(time.timeIntervalSince1970)))",
                     name: fileStore[url] as? String ?? URL(fileURLWithPath: filePath.path).lastPathComponent,
-                    id: idFromUrl(url: url) ?? ""
+                    id: idFromPodUrl(url: url) ?? ""
```
by definition, an id must be unique. If that's not the case, it's not an id.
Again, carried over from before, but as long as we're refactoring, we could maybe think a bit about it; maybe just a change of name would do.

timoteipalade replied:
It depends what is meant by the id field here. It could mean 2 things:

1. The id of the FileStats object. 
2. The id of the file/folder that we are creating a stat for. 

From what I understand the id here stands for the second option.

JJ commented on platform/ios/PolyPodApp/PodApi/PolyOut/FS/PolyOut+FS.swift:
```diff
@@ -107,7 +107,7 @@ extension PolyOut {
     func fileRead(url: String, options: [String: Any], completionHandler: @escaping (Any?, Error?) -> Void) {
         do {
-            guard let filePath = pathFromUrl(url: url) else {
+            guard let filePath = filesystemUrlFromPodUrl(url: url) else {
```
More carried over code. I would use uri, and not URL here, since we're talking about resource identifiers for files. Also, I would have to take a look at how `noSuchFile` works, but giving it an URL can be kind of confusing; an URI wouldn't be so confusing, but even so, the kind of error that's been thrown is not exactly noSuchFile, but "no file mapping for this URI", or maybe something more complicated.

JJ commented on platform/ios/PolyPodApp/PodApi/PolyOut/FS/PolyOut+FS.swift:
```diff
@@ -149,7 +149,7 @@ extension PolyOut {
                 return
             }
-            let targetUrl = pathFromId(id: url)
+            let targetUrl = filesystemUrlFromId(id: url)
```
The fact that you're using `url` as an `id` further supports the fact that it's an URI, not an URL.

timoteipalade replied:
Yap, I can use uri.

JJ commented on platform/ios/PolyPodApp/PodApi/PolyOut/FS/PolyOut+FS.swift:
```diff
@@ -43,17 +43,21 @@ extension PolyOut {
+    /// 2. Polypod url that starts with polypod:// and is how features access their resources in a platform independent way
     
-    func pathFromId(id: String) -> URL {
+    func filesystemUrlFromId(id: String) -> URL {
```
coming back here from down there. Please consider using `Uri` here and not `Url`. Also, shorten to `fsUri` or something like that.

JJ commented on platform/ios/PolyPodApp/PodApi/PolyOut/FS/PolyOut+FS.swift:
```diff
@@ -43,17 +43,21 @@ extension PolyOut {
-    }
+    /// There are 2 types of urls that are used
+    /// 1. Filesystem url that starts with file:// and points to a physical location on the device
+    /// 2. Polypod url that starts with polypod:// and is how features access their resources in a platform independent way
```
Sorry, I don't understand very well the context for this comment. Is it about the next function?

timoteipalade replied:
It is about the group of functions following it, related to mapping PolyPod urls to Filesystem urls.

JJ replied:
It would be better to try and find a better way to do it. One way that's used by some of us (notably @fhd) is to add a PR comment, so that it's clear for the reviewer, but does not leave a permanent footprint. Another option, not used by anyone, even more so in this Swift codebase, is to document the package, or class, or function. Since we don't have that, probably not the best way.

timoteipalade replied:
I can remove it. It was for me to clarify what is happening.

JJ reviewed (commented):
I'm a bit confused about where is exactly the code that allows to work with multiple archives. I would say it's in `handlePolyOutImportArchive`, but then it apparently deals only with two arguments (are they archives?) and not uniformly, with the second one being converted into a `destUrl`? (destination URL?). I'd appreciate if you clarified that.
As usual, some kind of unit tests (for instance, for URI mapping) would be helpful. If that's already covered, and it checks out OK, never mind that.

timoteipalade commented:
@JJ The idea is to use this API in the following way:
```
const destUrl = polyOut.importArchive(archiveUrl_1)
polyOut.importArchive(archiveUrl_2)
```
This way you can put multiple archives in the same place, which should be equivalent to importing multiple archives.

JJ commented:
> @JJ The idea is to use this API in the following way:
> 
> ```
> const destUrl = polyOut.importArchive(archiveUrl_1)
> polyOut.importArchive(archiveUrl_2, destUrl)
> ```
> 
> This way you can put multiple archives in the same place, which should be equivalent to importing multiple archives.

It would be probably a good thing if that was in a test, or documented somehow; it's going to be complicated otherwise to exactly know how to use it.

GhenadiePusca commented on platform/ios/PolyPodApp/PodApi/PolyOut/FS/PolyOut+FS.swift:
```diff
@@ -43,17 +43,21 @@ extension PolyOut {
     
-    private func idFromUrl(url: String) -> String? {
+    private func filesystemUrlFromPodUrl(url: String) -> URL? {
+        let id = idFromPodUrl(url: url)
+
+        return id != nil ? featureFilesPath().appendingPathComponent(id!) : nil
```
Just a suggestion, this could be simplified to something like this:
```suggestion
       idFromPodUrl(url: url).map(filesystemUrlFromId(id:))
```

timoteipalade commented on platform/ios/PolyPodApp/PodApi/PolyOut/FS/PolyOut+FS.swift:
```diff
@@ -208,10 +209,12 @@ extension PolyOut {
                     forKey: PolyOut.fsKey
                 ) as? [String:String?] ?? [:]
+                // TODO: With multiple archives support we can store more than one file in a single fileId.
+                // What should I do about the fileStore?
```
@GhenadiePusca. The fileStore can only deal with one filename. But multiple archives there can be multiple filenames stored for on fileId. 

I was thinking of changing fileStore to `[String: [String]]` instead of `[String: String?]`.

What do you think?

GhenadiePusca replied:
As far as I understood the changes made with this PR, your suggestion makes sense.

fhd replied:
I'm not Ghenadie, and maybe I don't get it, but why would we want to store multiple filenames for one field? I was thinking of logic like this:

1. See if `destUrl` was supplied, if not, generate a new UUID (you already have this above)
2. See if `destUrl` already exists, if not, create the directory
3. Extract the archive to `destUrl` - and now there's a design decision: We could overwrite identical files by default, we could throw an error is any file already exists (my preference for now), we could let the caller provide a conflict resolution strategy, even as a callback probably, but I think  this is overkill until we know what use cases we have.

So, however we do (3), I don't see a situation where we have more than a single path on disk per `destUrl` - what's called `newUrl` below.

timoteipalade replied:
@fhd Things work as you described above. And for (3) we overwrite the files. 

I think the point is another here. The fileStore object is a mapping where we keep track of which archives were extracted to which `destUrl`. And since we are able to extract more than one to the same `destUrl`, we need to modify it to support multiple files.

GhenadiePusca replied:
I guess it depends for what it needs to be used. 

Suppose we have one archive `my_facebook_data.zip` which is extracted somewhere in the folder `my_facebook_data`, and the FileStore would contain the `my_facebook_data` id for the destination url. 

Now if `my_facebook_data.zip` is split in, let's say, 3 archives. Then still `my_facebook_data_1.zip`, `my_facebook_data_2.zip`, `my_facebook_data_3.zip` would be extracted in one `my_facebook_data` folder. We can store the all 3 archive names for the same destination, for the means of knowing where a particular archive is unarchived.

But, looking again on the use cases, I am actually not sure if the 3 different names will be ever be used. I mean there is `removeArchive` API, which presumably removes the unarchived content. I don't know if it would be possible/needed to remove the content of a given sub-archive, like `my_facebook_data_2.zip`; as its unarchived content is now mixed with the content of the other 2 archives. So one would actually want to remove the content for all 3 archives in one go, which in this case would be `my_facebook_data` folder.

timoteipalade replied:
@GhenadiePusca @fhd Correct me if I am wrong, but the fileStore object does not make much sense. Why don't we just read the files at a destUrl from disk. I see that `readDir` already does that for the `url != ""` case. 

Am I missing something?

GhenadiePusca replied:
The name is used only one time in `stat` function. 
I think that we should be able to collapse multiple archive names in just one, which makes sense. I mean `my_facebook_data_1.zip`, `my_facebook_data_2.zip`, `my_facebook_data_3.zip` which would be extracted in just one folder, and I think that name of that folder should be stored in `fileStore`.

timoteipalade replied:
@GhenadiePusca It is also used in `readDir` at the end. And I find the use there pretty confusing. I will add a comment there.

timoteipalade replied:
Anyway, I will just make `fileStore` support multiple archive names, and then we can figure out how to do things properly when we move this logic to the core.

JJ replied:
> Well, in the Google case it is not about multiple archives then(if I understand correctly), as those can be perceived as separate/standalone entities, in the same manner as Facebook archive is different from Google archive.

Well, except they need to be included in a single `GoogleAccount`

JJ replied:
> > Is the intent documented somewhere?
> 
> The intent would be to show the correct/expected file name when `stat` function is called. But I guess it can be anything as it is just metadata.

But I wonder why we have no docs whatsoever in the Swift code base. Is that the usual practice? Not that we have a lot elsewhere; the polyPod, however, is relatively well documented. Going back to the original question, that intent would be documented in the `polyPod` documentation. You can generate it using typedoc (Not totally sure about that...)

GhenadiePusca replied:
> Well, except they need to be included in a single `GoogleAccount`

Makes sense, I imagine that Google feature probably would need to be able to read/modify separately each section, which means that each google archive should have different/unique/identifiable destination url, even though those are for the same google account. Based ob the current implementation in this PR, all of the archives are collapsed under the same url, and the info is lost.

Probably the proper usage from js should be:
```
const destUrlMaps = polyOut.importArchive(google_archive_maps, "GoogleAccount")
const destUrlMail = polyOut.importArchive(google_archive_mail, "GoogleAccount")
```
Where the two urls point to the specific archive, unarchived in the "GoogleAccount" folder.

And the archive url should look as such:
```swift
let id = UUID().string
let newUrl = PolyOut.fsPrefix + PolyOut.fsFilesRoot + "/" + destionationURL + "/" + id
```

@timoteipalade what do you think?

timoteipalade replied:
The files are split by features. See this in line 43:

```
func featureFilesPath() -> URL {
        let documentDirectory = FileManager.default.urls(for: .documentDirectory, in: .userDomainMask).first!.resolvingSymlinksInPath()
        return documentDirectory.appendingPathComponent(PolyOut.fsFilesRoot).appendingPathComponent(activeFeature!.id)
    }
    ```

GhenadiePusca replied:
Well then, that GoogleAccount folder can be put in the feature directory, still I am not sure if that wrapping in some `GoogleAccount` folder is needed at all.

timoteipalade replied:
The files are already segregated by feature, as in line 43, where the `activeFeature.id` is appended to the path. 

To me it makes sense that each archive goes into its own folder for google, since they are self contained.

But what about other features where the archives would need to be in the same place? 

The change in this PR would support both. You can choose whether you put all archives in one place, or not.

timoteipalade replied:
We need to bring up this thing about where we store the data, and how it should be structured in the next weekly dev, and document it somewhere.

timoteipalade replied:
> Well then, that GoogleAccount folder can be put in the feature directory, still I am not sure if that wrapping in some `GoogleAccount` folder is needed at all.

I don't think it is.

GhenadiePusca replied:
Ok, so destinationUrl argument will be only helpful when needing to unarchive at the same location.
 I am somewhat confused, as the trigger for this PR was Google exporting multiple archives, but from the looks of it, it was already supported with the old implementation, or I am completely missing something this Friday afternoon 😄?

timoteipalade replied:
Say you have 2 archives and no destinationUrl. How would you import them in the same place?

GhenadiePusca replied:
I understand, but does Google import need this? As JJ mentioned, each google archive is standalone. 
For Facebook example, when a big zip is split in multiple ones for the means of space, yes, importing to the same location is what is needed, and this PR achieves it.

timoteipalade replied:
According to the ticket, we want to put all archives in one place: https://jira.polypoly.eu/browse/PROD4POD-1523

GhenadiePusca replied:
I struggle to understand the need, as it seems that all of the google archives are standalone, do not need merging, and would need probably separate identification.

Anyway, the PR does the job. Just that in my head, Google import would work also with the old implementation.

timoteipalade replied:
Yeah, you have a valid point in my opinion. @fhd are we missing something?

timoteipalade commented:
@JJ There will be tests when this stuff is finished, either manual or automatic, or some sort of documentation. Right now I just want to know if I am on the right path with this ticket.

JJ commented:
> > It would be probably a good thing if that was in a test, or documented somehow; it's going to be complicated otherwise to exactly know how to use it.
> 
> Once we have the _test_ feature :)

Sooner or later, someone's gonna have to do that...

fhd commented on platform/ios/PolyPodApp/PodApi/PolyOut/FS/PolyOut+FS.swift:
```diff
@@ -43,17 +43,17 @@ extension PolyOut {
     }
     
-    private func idFromUrl(url: String) -> String? {
+    private func idFromPodUrl(_ url: String) -> String? {
```
Since you're renaming anyway, is this really appropriate to be called `id`? Isn't it more like.... `relativePath`? But I can't say I _fully_ understand this code and all these conversions :D

timoteipalade replied:
The result of this method is a `UUID`. I think this is okay.

fhd reviewed (commented):
Looks like you're moving in the right direction! I'm just confused about the code in general (not your fault, but maybe you can shed a light on the questions I have), and about why you don't just extract the contents of an additional archive into the existing path if it already exists, seems like the no-brainer approach to me.

timoteipalade commented on platform/ios/PolyPodApp/PodApi/PolyOut/FS/PolyOut+FS.swift:
```diff
@@ -170,7 +167,7 @@ extension PolyOut {
         // actually exist - for now we just ignore them in readDir, but it
         // might be smarter to clean fileStore automatically at some point.
         let storedFiles = fileStore.keys.filter { key in
```
@GhenadiePusca this is the place where the use of fileStore does not make much sense to me. Why give the contents of the folders in the case of `url != ""` and give the names of the archives in the other case? 

It is strange

timoteipalade commented on platform/ios/PolyPodApp/PodApi/PolyOut/FS/PolyOut+FS.swift:
```diff
@@ -149,7 +145,10 @@ extension PolyOut {
                 return
             }
-            let targetUrl = pathFromId(id: url)
+            guard let targetUrl = fsUriFromPodUrl(url) else {
```
Yeah, pretty much it.

timoteipalade commented:
> Besides the conflict, there seems to be a warning now. Can you please check that out?

@JJ Where do you see the warning? When executing `npm run build`?

JJ commented on platform/feature-api/communication/src/remote-pod/remote.ts:
```diff
@@ -240,8 +240,8 @@ export class RemoteClientPod implements Pod {
 
-            importArchive(url: string): Promise<string> {
-                return rpcClient.polyOut().importArchive(url)();
+            importArchive(url: string, destUrl?: string): Promise<string> {
```
One of the issues why I want to separate core from other changes, is to check that they are properly tested. This is a part of the code base that has a high coverage, and now we've simply created a different use case, with no tests to go with it. Since it's an optional argument, all tests will pass, of course, and the only real code there is is this one; still, all tests will be set with a null destUrl, which is not what we want.
Also, different use cases have been created. From your comments elsewhere, I think that the intention is to make the first call without a destUrl; the rest *with* them. So what happens if we do calls in a different sequence? Will that work?

timoteipalade replied:
I will add tests. Calls in a different sequence are possible, but don't make sense, since the first call creates the `destUrl` for the rest of the calls.

JJ commented:
> > Besides the conflict, there seems to be a warning now. Can you please check that out?
> 
> @JJ Where do you see the warning? When executing `npm run build`?

It's right there on the "files" tab. Pops out like a sore thumb. I've also commented how to get rid of it.

JJ commented:
Restarting tests. Lately MacOS tests seem to be failing for no good reason.

timoteipalade commented on platform/feature-api/api/pod-api/src/default.ts:
```diff
@@ -105,7 +105,7 @@ export class DefaultPod implements Pod {
             }
 
-            importArchive(url: string): Promise<string> {
+            importArchive(url: string, destUrl?: string): Promise<string> {
```
Got it! Thanks!

timoteipalade commented on platform/feature-api/api/pod-api/src/fs.ts:
```diff
@@ -31,7 +31,7 @@ export interface FS {
     stat(path: string): Promise<Stats>;
     readdir(path: string): Promise<string[]>;
-    importArchive(url: string): Promise<string>;
+    importArchive(url: string, destUrl?: string): Promise<string>;
```
I am not sure I get the question about the mappings. Can you elaborate a bit?

JJ replied:
Edited to clarify.

timoteipalade replied:
Let me define a couple of terms:

1. Polypod URL - a url used by a feature to access resources. This needs to be platform independent, and yield the same result independent of the platform. 
2. Client URL - a url that the client uses to access a resource for that a feature requested using a Polypod URL. 

Mapping from Polypod URL to Client URL is the client's responsibility. For example on iOS we use the filesystem to store stuff, but with JS we use IndexedDB, which works in a different way. 

Now coming to the questions. I will write how I understand your questions. Let me know if it is correct. 

> While before deleting fileId might tell the client to delete the source, now can that mapping be reversed?

Is the mapping between PolyPod URL and Client URL isomorphic (does it go both ways)? 

> Talking about mappings, not totally clear either what's the mapping from the fact that we are adding multiple archive functionality to the need of this additional destUrl. Can you please clarify?

How is `destUrl` related to / necessary for supporting multiple archives? 

Did I get it right?

JJ replied:
> > While before deleting fileId might tell the client to delete the source, now can that mapping be reversed?
> 
>     1. Is the mapping between PolyPod URL and Client URL isomorphic (does it go both ways)?

Well, we now have also a destURL. It simply can't be isomorphic between them. 
> 
> 
> > Talking about mappings, not totally clear either what's the mapping from the fact that we are adding multiple archive functionality to the need of this additional destUrl. Can you please clarify?
> 
>     2. How is `destUrl` related to / necessary for supporting multiple archives?

Pretty much, yes. Those are my doubts.

timoteipalade replied:
JJ and I had a conversation and we discussed question 2 above. 

At the end of the conversation it was decided that I should add tests, and document importArchive.

JJ commented on platform/podjs/src/browserPod.ts:
```diff
@@ -15,6 +15,7 @@ import * as RDFString from "rdf-string";
 import * as zip from "@zip.js/zip.js";
 import endpointsJson from "../../../../polyPod-config/endpoints.json";
 import { Manifest, readManifest } from "./manifest";
+import "cross-fetch/polyfill";
```
Do you need this *and* babel?

JJ commented on platform/podjs/src/browserPod.ts:
```diff
@@ -330,32 +328,83 @@ class IDBPolyOut implements PolyOut {
+    /// destUrl should be the same as zipId
+    async importArchive(url: string, destUrl?: string): Promise<string> {
+        const { fileName } = FileUrl.fromUrl(url);
+        const buffer = await (await fetch(url)).arrayBuffer();
```
OK, you need `cross-fetch` to do this, I guess; since this was not tested before at the unit level, it didn't fail.
You probably made a good choice; however, did you consider node-fetch? It's what's used in some other modules (`remote-pod` IIRC). When we use the same module in two or more places, we "hoist" it to the root so that it's installed only once.

timoteipalade replied:
I looked at node-fetch, but I need something that works both with node and browsers, without me having to change which function I call based on the platform.

JJ reviewed (commented):
Just comments, since this is a draft. Major issue: new binary file on the repo. Let me check about the zip generation, however.

timoteipalade commented on platform/podjs/src/tests/browserPod.test.ts:
```diff
@@ -0,0 +1,15 @@
+test("Simple Archive Import", async () => {
+    const polyOut = new IDBPolyOut();
+    try {
+        const zipPath =
```
I will remove the zip file. I forgot to remove it.

timoteipalade commented on platform/podjs/package.json:
```diff
@@ -18,17 +18,23 @@
   },
   "devDependencies": {
+    "@babel/plugin-transform-runtime": "^7.17.10",
+    "@babel/preset-env": "^7.17.10",
```
Yap, I will remove it tomorrow

timoteipalade commented on platform/podjs/cypress.json:
```diff
@@ -1,4 +1,5 @@
     "pluginsFile": false,
-    "video": false
+    "video": false,
+    "chromeWebSecurity": false
```
Yeah, I did not want to disable `chromeWebSecurity` but when I try to fetch the zip over https I get this error `TypeError: Failed to fetch`. I found something about it here: https://github.com/cypress-io/cypress/issues/2047, but it does not seem to be figured out. Do you guys know what is the issue?

JJ replied:
Well, fetching stuff from outside the repo is tricky. It adds to testing time, for instance. It might clash with sandboxing.

JJ replied:
Do you still need it?

timoteipalade commented:
Update: iOS and pod.js have the functionality implemented and tested as well. 

I am working on Android rn.

timoteipalade commented:
Okay guys, it is ready for review! 

Functionality is implemented and tested for all three platforms. 

There's one issue that bothers me. Each platform returns a different format for the result of `importArchive`. 

JS returns: "polypod://<uuid>"
iOS returns: "polypod://FeatureFiles/<uuid>"
Android returns: "<uuid>"

JJ commented:
> Okay guys, it is ready for review!
> 
> Functionality is implemented and tested for all three platforms.
> 
> There's one issue that bothers me. Each platform returns a different format for the result of `importArchive`.
> 
> JS returns: `polypod://<uuid>` iOS returns: `polypod://FeatureFiles/<uuid>` Android returns: `<uuid>`

But is that internal to every platform? Or will it be going back and forth between the different platforms?

timoteipalade commented:
> > Okay guys, it is ready for review!
> > Functionality is implemented and tested for all three platforms.
> > There's one issue that bothers me. Each platform returns a different format for the result of `importArchive`.
> > JS returns: `polypod://<uuid>` iOS returns: `polypod://FeatureFiles/<uuid>` Android returns: `<uuid>`
> 
> But is that internal to every platform? Or will it be going back and forth between the different platforms?

It is not internal. These values make it back to the features. So they might end up going back and forward between different platforms.

GhenadiePusca commented on platform/android/app/src/main/kotlin/coop/polypoly/polypod/polyOut/PolyOut.kt:
```diff
@@ -154,24 +157,26 @@ open class PolyOut(
-                    val newId = UUID.randomUUID().toString()
                     val fs = Preferences.getFileSystem(context).toMutableMap()
-                    fs[fsPrefix + newId] = fileName
+                    fs[fsPrefix + zipId] = fileName
```
Why this is different from iOS where a list of filenames is stored for the same path?

timoteipalade replied:
I missed that. Good catch.

GhenadiePusca commented on platform/ios/PolyPodApp/PolyPodTests/PolyOutTests.swift:
```diff
@@ -146,4 +146,54 @@ class PolyOutTests: XCTestCase {
+        
+        let expectation = XCTestExpectation(description: "Exp")
+        polyOut.importArchive(url: url.absoluteString, destUrl: nil) { newUrl in
+            XCTAssertTrue(newUrl != nil && newUrl != "", "newUrl is nil or empty")
```
Nit: You should be consistent with how you check for emptiness, I see you did use 3 ways in this test function :). `isEmpty` should be preferred to `!= ""` or `.count > 0`.

timoteipalade replied:
#diversity 😛

GhenadiePusca commented on platform/ios/PolyPodApp/PolyPodTests/PolyOutTests.swift:
```diff
@@ -146,4 +146,54 @@ class PolyOutTests: XCTestCase {
+        let url1 = bundle.url(forResource: "multipleZips1", withExtension: "zip")!
+        let url2 = bundle.url(forResource: "multipleZips2", withExtension: "zip")!
+        
+        let expectation = XCTestExpectation(description: "Exp")
+        polyOut.importArchive(url: url1.absoluteString, destUrl: nil) { newUrl1 in
+            XCTAssertTrue(newUrl1 != nil && newUrl1 != "", "newUrl1 is nil or empty")
+            
+            polyOut.importArchive(url: url2.absoluteString, destUrl: newUrl1!) { newUrl2 in
+                XCTAssertTrue(newUrl2 != nil && newUrl2 != "", "newUrl2 is nil or empty")
+                XCTAssertTrue(newUrl1 == newUrl2)
+                
+                polyOut.readDir(url: newUrl1!) { stuff, error in
+                    XCTAssertNil(error, "error is not nil")
+                    XCTAssertTrue(stuff != nil && !stuff!.isEmpty, "no files were found")
+                    XCTAssertTrue(stuff!.filter{ $0["path"] == "multipleZips1/file1.rtf" }.count > 0, "file1  not found")
+                    XCTAssertTrue(stuff!.filter{ $0["path"] == "multipleZips2/file2.rtf" }.count > 0, "file2  not found")
+                    expectation.fulfill()
+                }
+            }
+        }
+        wait(for: [expectation], timeout: 20.0)
```
Nit:
Tests will be easier to read if assertions are not done in completion blocks. I mean something like this:
```swift
let newURL1 = importArchive(url: url1.absoluteString, polyOut: polyOut)
XCTAssertTrue(newUrl1 != nil && newUrl1 != "", "newUrl1 is nil or empty")
let newURL1 = importArchive(url: url2.absoluteString, polyOut: polyOut)
XCTAssertTrue(newUrl2 != nil && newUrl2 != "", "newUrl2 is nil or empty")
```

timoteipalade commented on platform/android/app/src/main/kotlin/coop/polypoly/polypod/features/Feature.kt:
```diff
@@ -6,25 +6,36 @@ import android.graphics.Color
 import android.util.DisplayMetrics
 import java.util.zip.ZipFile
 
+interface FeatureInterface {
```
Yeah, the constructor thing does not work. A second constructor needs to reference the first (primary) one, and `findUrl` requires a manifest, which I don't want to reconstruct or pass to the second constructor, since it is such a pain to deal with.

fhd replied:
Not a huge fan of having something with `Interface` in the name either - maybe the interface could be `Feature` and the concrete implementation below `ZipFeature` or something? Anything _other_ than adding stuff like `Interface` or `Impl` (yuck) seems better :D

GhenadiePusca replied:
> Yeah, the constructor thing does not work. A second constructor needs to reference the first (primary) one, and `findUrl` requires a manifest, which I don't want to reconstruct or pass to the second constructor, since it is such a pain to deal with.

You can have the constructor which initialises all the fields to be the primary constructor, and the current constructor will be secondary one. Also from the code, `findURL` just uses `Manifest` to access `links`, which could be sent at initialisation.

timoteipalade commented on platform/android/app/src/main/kotlin/coop/polypoly/polypod/ZipTools.kt:
```diff
@@ -22,15 +22,23 @@ class ZipTools {
+                        file.parentFile.mkdirs()
+                    }
+
+                    if (!entry.isDirectory) {
```
It can, but I wanted to keep to logical areas separate. First logical area creates the directories. The second one does the extraction.

timoteipalade commented on platform/android/app/src/test/kotlin/coop/polypoly/polypod/polyout/PolyOutTest.kt:
```diff
@@ -0,0 +1,298 @@
+    }
+
+    private var polyOut: PolyOut = PolyOut(
+        context = InstrumentationRegistry.getInstrumentation().targetContext
```
Nope, context is not initialised yet.

GhenadiePusca replied:
True, but Kotlin also has the concept of lazy variables.

JJ replied:
Is this the private var you're defining below? Is there any reason why it's not defined above?

timoteipalade replied:
Easier to read the code with all the mocked classes above.

JJ replied:
Right, but if you finally follow @GhenadiePusca suggestion it would make sense in order to understand there's a local `context` variable that's assigned to that argument.

timoteipalade commented on platform/android/app/src/main/kotlin/coop/polypoly/polypod/polyOut/PolyOut.kt:
```diff
@@ -154,24 +157,26 @@ open class PolyOut(
             }
         }.await()
-        return null // TODO: Return the expected value
+        return zipId
```
I wanted to bring attention to this issue. I am not sure what should be returned. It seems a uuid is enough. 

What should be returned and why? That's something I haven't managed to answer yet, if there is an answer.

GhenadiePusca replied:
Well on iOS import archive will return the following: `polypod://FeatureFiles/<uuid>`, while on Android just `<uuid>`, shouldn't it be the same on both platforms?

JJ replied:
Maybe a candidate for shared code? (later, of course)

fhd replied:
Actually yes, it should return the same thing, a URL, not a path without protocol, so nevermind what I said :D

timoteipalade replied:
I made all platforms return the same thing :)

timoteipalade commented on platform/podjs/cypress/integration/pod_spec.js:
```diff
@@ -4,3 +4,46 @@ import "../../dist/pod.js";
+describe("Import Archive", () => {
+    it("only one archive", async () => {
+        const zipPath =
+            "https://github.com/timoteipalade/test-files/raw/main/testZip.zip";
```
😆 Yeah, I am not sure what to do about those. It was a desperate move. I had a local server for them, but node does not allow for self signed certificates (https).

fhd replied:
That's pretty nasty indeed :D Some ideas:

1. Maybe a data URL works :D
2. Maybe an absolute file URL would work? (Probably worth inspecting what `polyNav.pickFile` returns in Chrome - a file URL?)
3. Cypress already runs a server i.e., maybe you could sneak this one into the bundle?

timoteipalade replied:
I went with 3.

fhd reviewed (approved):
Took a _super_ high level look, since I think we have enough cooks on this one for the details, but seems pretty spot on!

fhd commented on platform/podjs/cypress/integration/pod_spec.js:
```diff
@@ -4,3 +4,46 @@ import "../../dist/pod.js";
     podSpec(window.pod, "");
 });
+
+describe("Import Archive", () => {
```
The `pod_spec.js` file is meant for running the spec tests defined within _feature-api_ - so I think this warrants a separate file.

JJ commented on platform/android/app/src/test/kotlin/coop/polypoly/polypod/polyout/PolyOutTest.kt:
```diff
@@ -0,0 +1,298 @@
+    }
+
+    @Suppress("TooManyFunctions")
+    class FakeKeyStore : KeyStoreSpi() {
```
That would also avoid the `TooManyFunctions` warning.

JJ commented on platform/android/app/src/test/kotlin/coop/polypoly/polypod/polyout/PolyOutTest.kt:
```diff
@@ -0,0 +1,298 @@
+        )
+    }
+
+    @After
```
Is this mandatory even if it does nothing?

JJ commented on platform/android/app/src/test/kotlin/coop/polypoly/polypod/polyout/PolyOutTest.kt:
```diff
@@ -0,0 +1,298 @@
+        val url1 = InstrumentationRegistry
+            .getInstrumentation()
+            .javaClass.classLoader!!
+            .getResource("multipleZips1.zip")
```
Same as said above, extract up to here to a variable and reuse it in the assignments below.

JJ commented on platform/android/app/src/test/kotlin/coop/polypoly/polypod/polyout/PolyOutTest.kt:
```diff
@@ -0,0 +1,298 @@
+        val url2 = InstrumentationRegistry
+            .getInstrumentation()
+            .javaClass.classLoader!!
+            .getResource("multipleZips2.zip")
```
Ditto, and also you could maybe use a loop over the two literals to DRY.

JJ commented on platform/android/app/src/test/kotlin/coop/polypoly/polypod/polyout/PolyOutTest.kt:
```diff
@@ -0,0 +1,298 @@
+        Truth.assertThat(hasFile1).isTrue()
+        Truth.assertThat(hasFile2).isTrue()
+    }
+}
```
(This is for below) I don't really like binaries in the repo. Correct way of doing it would be uploading them to a release, or generating them on the fly during the setup phase.

JJ commented on platform/podjs/cypress/integration/pod_spec.js:
```diff
@@ -4,3 +4,46 @@ import "../../dist/pod.js";
+
+    it("multiple archives", async () => {
+        const zipPath1 =
+            "https://github.com/timoteipalade/test-files/raw/main/multipleZips1.zip";
```
OK, but they are now in the repo, right?

JJ reviewed (approved):
I'm going to go green here, even if there's a thing I particularly don't like: binary files in the repo, even more so *repeated* binary files in the repo, and even worse, *repeated* binary files in the repo and *outside* the repository.
However, generating them, which would be, I guess, the best alternative, would imply repeating *the code* in the three different languages. Creating shared code is obviously outside the scope of this PR. So anything that avoids replication would be OK, (but also OK if that's left for later)

timoteipalade commented:
Okay, I think we are ready to merge this! 

What I addressed:

1. For cypress, we now load the zips from the repo. When the test runs on a machine it no longer needs to make requests, eliminating flakiness from the internet connection. 
2. All platforms return the same thing for importArchive
3. I forgot to address storing multiple filenames for the same id on Android. I took care of it.
4. I cleaned the tests.

GhenadiePusca commented on platform/ios/PolyPodApp/PolyPod.xcodeproj/project.pbxproj:
```diff
@@ -104,6 +104,9 @@
 		906E406C27A82C7F009015B7 /* ActivityViewController.swift in Sources */ = {isa = PBXBuildFile; fileRef = 906E406B27A82C7F009015B7 /* ActivityViewController.swift */; };
 		90E4088227D78E4E006AD362 /* DataProtection.swift in Sources */ = {isa = PBXBuildFile; fileRef = 90E4088127D78E4E006AD362 /* DataProtection.swift */; };
 		90E4088327D79382006AD362 /* DataProtection.swift in Sources */ = {isa = PBXBuildFile; fileRef = 90E4088127D78E4E006AD362 /* DataProtection.swift */; };
+		9C856C20282E5126004DF92F /* multipleZips1.zip in Resources */ = {isa = PBXBuildFile; fileRef = 9C856C1D282E5126004DF92F /* multipleZips1.zip */; };
```
Probably we could create a zip programatically, but this is good enough for now.

snoack commented:
Why was `FileInfo.blob: Blob` changed to `FileInfo.buffer : ArrayBuffer` in *podjs*? Using a `Blob` has the advantage that the file contents don't need to be read from disk when only the metadata need to be accessed.

timoteipalade commented:
> Why was `FileInfo.blob : Blob` changed to `FileInfo.buffer : ArrayBuffer` in _podjs_? Using a `Blob` has the advantage that the file contents don't need to be read from disk when only the metadata need to be accessed.

Because I could not test it. The fake-indexedDB implementation, used for tests, does not support saving Blobs. 

I was at the end of my wits with that issue, so instead of having untested code, I decided to go with the more inefficient solution. 

But since I am not using Jest and fake-indexedDB anymore, using ArrayBuffer is unnecessary. Shoot. I missed this. 

Very good catch. I will revert to using Blob.

timoteipalade commented:
> Why was `FileInfo.blob : Blob` changed to `FileInfo.buffer : ArrayBuffer` in _podjs_? Using a `Blob` has the advantage that the file contents don't need to be read from disk when only the metadata need to be accessed.

@snoack PR opened for this here: https://github.com/polypoly-eu/polyPod/pull/905

JJ commented:
> Why was `FileInfo.blob : Blob` changed to `FileInfo.buffer : ArrayBuffer` in _podjs_? Using a `Blob` has the advantage that the file contents don't need to be read from disk when only the metadata need to be accessed.

That would have been an excellent target for a test. If that's a behavior you need enforced, the best way is to write a test that checks that and that will fail if the signature is changed in any way, as in this case.

timoteipalade commented:
> > Why was `FileInfo.blob : Blob` changed to `FileInfo.buffer : ArrayBuffer` in _podjs_? Using a `Blob` has the advantage that the file contents don't need to be read from disk when only the metadata need to be accessed.
> 
> That would have been an excellent target for a test. If that's a behavior you need enforced, the best way is to write a test that checks that and that will fail if the signature is changed in any way, as in this case.

A bit overkill if you ask me 😆

timoteipalade commented on platform/podjs/src/browserPod.ts:
```diff
@@ -253,31 +253,40 @@ interface File {
         return new Promise((resolve, reject) => {
             const tx = db.transaction([OBJECT_STORE_POLY_OUT], "readonly");
-            const request = tx.objectStore(OBJECT_STORE_POLY_OUT).get(id);
+            const request = tx.objectStore(OBJECT_STORE_POLY_OUT).get(zipId);
```
I have changed the way I save into the db. I now pass the zipId as a second argument when I write in `importArchive`.

snoack replied:
I see. How about using an index instead of maintaining an array of files in storage for each id? I had a go here: #906

timoteipalade commented on platform/podjs/src/browserPod.ts:
```diff
@@ -297,23 +306,9 @@ class IDBPolyOut implements PolyOut {
-                    false
-                );
-            },
-        };
```
🤔 I might be mistaken about this. I will check it tomorrow.

snoack replied:
Apparently that is another reason why the `google` and `facebookImport` feature is broken in *pod.js* since this change landed. So yeah, it wasn't dead code.

snoack commented on platform/podjs/src/browserPod.ts:
```diff
@@ -358,32 +355,83 @@ class IDBPolyOut implements PolyOut {
+    /// destUrl should be the same as zipId
+    async importArchive(url: string, destUrl?: string): Promise<string> {
+        const { fileName } = FileUrl.fromUrl(url);
+        const buffer = await (await fetch(url)).arrayBuffer();
```
This is a regression. Since this change got merged importing an archive in *podjs* fails with follwing error being logged in the console:

    pod.js:9355 GET data:application/zip;base64,UEsDBBQAAAAAAAD43p0AAAAAAAAAAAAAAAAaAAAAZ29vZ2x…1sUEsFBgAAAACaAJoA9UoAAAE5cAAAAA==/google-fake-malcomgerlach-2022-3-22.zip net::ERR_INVALID_URL

Note that previously we called `fetch(FileUrl.fromUrl(url).data)` but now you are passing the `url` unchanged to `fetch()`.
